crates/iceberg/src/avro/schema.rs

```diff
@@ -0,0 +1,917 @@
+        }
+    }
+
+    fn primitive(&mut self, p: &PrimitiveType) -> Result<Either<AvroSchema, AvroRecordField>> {
+        let avro_schema = match p {
+            PrimitiveType::Boolean => AvroSchema::Boolean,
+            PrimitiveType::Int => AvroSchema::Int,
+            PrimitiveType::Long => AvroSchema::Long,
+            PrimitiveType::Float => AvroSchema::Float,
+            PrimitiveType::Double => AvroSchema::Double,
+            PrimitiveType::Date => AvroSchema::Date,
+            PrimitiveType::Time => AvroSchema::TimeMicros,
+            PrimitiveType::Timestamp => AvroSchema::TimestampMicros,
+            PrimitiveType::Timestamptz => AvroSchema::TimestampMicros,
+            PrimitiveType::String => AvroSchema::String,
+            PrimitiveType::Uuid => AvroSchema::Uuid,
+            PrimitiveType::Fixed(len) => avro_fixed_schema((*len) as usize)?,
+            PrimitiveType::Binary => AvroSchema::Bytes,
+            PrimitiveType::Decimal { precision, scale } => {
+                avro_decimal_schema(*precision as usize, *scale as usize)?
+            }
+        };
+        Ok(Either::Left(avro_schema))
+    }
+}
+
+/// Converting iceberg schema to avro schema.
+pub(crate) fn schema_to_avro_schema(name: impl ToString, schema: &Schema) -> Result<AvroSchema> {
+    let mut converter = SchemaToAvroSchema {
+        schema: name.to_string(),
+    };
+
+    visit_schema(schema, &mut converter).map(Either::unwrap_left)
+}
+
+pub(crate) fn avro_fixed_schema(len: usize) -> Result<AvroSchema> {
+    Ok(AvroSchema::Fixed(FixedSchema {
+        name: Name::new(format!("fixed_{len}").as_str())?,
+        aliases: None,
+        doc: None,
+        size: len,
+        attributes: Default::default(),
+    }))
+}
+
+pub(crate) fn avro_decimal_schema(precision: usize, scale: usize) -> Result<AvroSchema> {
+    Ok(AvroSchema::Decimal(DecimalSchema {
+        precision,
+        scale,
+        inner: Box::new(avro_fixed_schema(DECIMAL_LENGTH)?),
+    }))
+}
+
+fn avro_optional(avro_schema: AvroSchema) -> Result<AvroSchema> {
+    Ok(AvroSchema::Union(UnionSchema::new(vec![
+        AvroSchema::Null,
+        avro_schema,
+    ])?))
+}
+
+fn is_avro_optional(avro_schema: &AvroSchema) -> bool {
+    match avro_schema {
+        AvroSchema::Union(union) => union.is_nullable(),
+        _ => false,
+    }
+}
+
+/// Post order avro schema visitor.
+pub trait AvroSchemaVisitor {
+    type T;
+
+    fn record(&mut self, record: &RecordSchema, fields: Vec<Self::T>) -> Result<Self::T>;
+
+    fn union(&mut self, union: &UnionSchema, options: Vec<Self::T>) -> Result<Self::T>;
+
+    fn array(&mut self, array: &AvroSchema, item: Self::T) -> Result<Self::T>;
+    fn map(&mut self, map: &AvroSchema, value: Self::T) -> Result<Self::T>;
+
+    fn primitive(&mut self, schema: &AvroSchema) -> Result<Self::T>;
+}
+
+/// Visit avro schema in post order visitor.
+pub fn visit<V: AvroSchemaVisitor>(schema: &AvroSchema, visitor: &mut V) -> Result<V::T> {
```

Comment:
I wonder what the benefit of the visitor pattern is compared to using pattern matching as in the following code:

```rust
/// Converts avro schema to iceberg schema.
pub(crate) fn avro_schema_to_schema(avro_schema: &AvroSchema) -> Result<Schema> {
    if let AvroSchema::Record(_) = avro_schema {
        let typ = avro_type_to_type(avro_schema, 0)?;
        if let Some(Type::Struct(s)) = typ {
            Schema::builder()
                .with_fields(s.fields().iter().cloned())
                .build()
        } else {
            Err(Error::new(
                ErrorKind::Unexpected,
                format!("Expected to convert avro record schema to struct type, but {typ}"),
            ))
        }
    } else {
        Err(Error::new(
            ErrorKind::DataInvalid,
            "Can't convert non record avro schema to iceberg schema: {avro_schema}",
        ))
    }
}

fn avro_type_to_type(avro: &AvroSchema, next_field_id: i32) -> Result<Option<Type>> {
    let typ = match avro {
        AvroSchema::Decimal(decimal) => {
            Type::decimal(decimal.precision as u32, decimal.scale as u32)?
        }
        AvroSchema::Date => Type::Primitive(PrimitiveType::Date),
        AvroSchema::TimeMicros => Type::Primitive(PrimitiveType::Time),
        AvroSchema::TimestampMicros => Type::Primitive(PrimitiveType::Timestamp),
        AvroSchema::Uuid => Type::Primitive(PrimitiveType::Uuid),
        AvroSchema::Boolean => Type::Primitive(PrimitiveType::Boolean),
        AvroSchema::Int => Type::Primitive(PrimitiveType::Int),
        AvroSchema::Long => Type::Primitive(PrimitiveType::Long),
        AvroSchema::Float => Type::Primitive(PrimitiveType::Float),
        AvroSchema::Double => Type::Primitive(PrimitiveType::Double),
        AvroSchema::String | AvroSchema::Enum(_) => Type::Primitive(PrimitiveType::String),
        AvroSchema::Fixed(fixed) => Type::Primitive(PrimitiveType::Fixed(fixed.size as u64)),
        AvroSchema::Bytes => Type::Primitive(PrimitiveType::Binary),
        AvroSchema::Null => return Ok(None),
        AvroSchema::Record(record) => {
            let field_types = record
                .fields
                .iter()
                .map(|x| avro_type_to_type(&x.schema, next_field_id))
                .collect::<Result<Vec<Option<Type>>>>()?;
            let mut fields = Vec::with_capacity(field_types.len());
            for (avro_field, typ) in record.fields.iter().zip_eq(field_types) {
                let field_id = avro_field
                    .custom_attributes
                    .get(FILED_ID_PROP)
                    .and_then(Value::as_i64)
                    .ok_or_else(|| {
                        Error::new(
                            ErrorKind::DataInvalid,
                            format!("Can't convert field, missing field id: {avro_field:?}"),
                        )
                    })?;

                let optional = is_avro_optional(&avro_field.schema);

                let mut field = if optional {
                    NestedField::optional(field_id as i32, &avro_field.name, typ.unwrap())
                } else {
                    NestedField::required(field_id as i32, &avro_field.name, typ.unwrap())
                };

                if let Some(doc) = &avro_field.doc {
                    field = field.with_doc(doc);
                }

                fields.push(field.into());
            }

            Ok(Some(Type::Struct(StructType::new(fields))))
        }
        AvroSchema::Array(array) => {
            let item = avro_type_to_type(array, next_field_id)?;
            if let AvroSchema::Array(item_schema) = array.as_ref() {
                let element_field = NestedField::list_element(
                    next_field_id + 1,
                    item.unwrap(),
                    !is_avro_optional(item_schema),
                )
                .into();
                Ok(Some(Type::List(ListType { element_field })))
            } else {
                Err(Error::new(
                    ErrorKind::Unexpected,
                    "Expected avro array schema, but {array}",
                ))
            }
        }
    };
    Ok(Some(typ))
}
```

For me this is much clearer what happens.

Reply:
IMO the benefit of the vistor is to reuse it for coverting or collect info in schema (They represent different business logic). If it only used in `avro schema to schema`, maybe recursive function is enough.

Reply:
The problem is whether we will need to process avro schema in other places. If not, I think using recursive + pattern matching is good enough.

Reply:
We can also use the visitor pattern here. I'm just generally trying to be cautious when porting OO code to rust not to introduce complexity that is not necessary in Rust.

Reply:
Let's go with the visitor pattern. I just want to make a couple of points for completeness sake.

From [wikipedia](https://en.wikipedia.org/wiki/Visitor_pattern#VisitorPattern): 

Programming languages with [sum types](https://en.wikipedia.org/wiki/Tagged_union) and [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) obviate many of the benefits of the visitor pattern, as the visitor class is able to both easily branch on the type of the object and generate a compiler error if a new object type is defined which the visitor does not yet handle.

The benefits of the visitor pattern don't apply to our case:

1. Visitor allows same algorithm over different types

We are always visiting the same type, even with a read schema and a write schema.

2. Visitor allows dynamic double dispatch

Without any default implementations on the visitor trait, all trait methods have to be implemented for a new type which is equal to pattern matching. There would be a benefit if you have many methods with default implementations because then you could save implementation effort for new types.

Reply:
> Converting the schema might be a bit simple for the visitor, but later on, when you start pruning schema (or for Avro, having a read schema that differs from the write schema), it also gives this concept of divide and conquer where you can easily isolate the different cases.

This is also my concern. I prefer to keep visitor pattern to see if we have more features to implement on avro schema.